src/request.rs

```diff
@@ -42,6 +43,144 @@ impl AppendOnlyData {
     pub fn tag(&self) -> u64 {
         self.address().tag()
     }
+
+    pub fn permissions_index(&self) -> u64 {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.permissions_index(),
+            AppendOnlyData::PubUnseq(data) => data.permissions_index(),
+            AppendOnlyData::UnpubSeq(data) => data.permissions_index(),
+            AppendOnlyData::UnpubUnseq(data) => data.permissions_index(),
+        }
+    }
+
+    pub fn owners_index(&self) -> u64 {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.owners_index(),
+            AppendOnlyData::PubUnseq(data) => data.owners_index(),
+            AppendOnlyData::UnpubSeq(data) => data.owners_index(),
+            AppendOnlyData::UnpubUnseq(data) => data.owners_index(),
+        }
+    }
+
+    pub fn in_range(&self, start: ADataIndex, end: ADataIndex) -> Option<Vec<(Vec<u8>, Vec<u8>)>> {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.in_range(start, end),
+            AppendOnlyData::PubUnseq(data) => data.in_range(start, end),
+            AppendOnlyData::UnpubSeq(data) => data.in_range(start, end),
+            AppendOnlyData::UnpubUnseq(data) => data.in_range(start, end),
+        }
+    }
+
+    pub fn indices(&self) -> Result<ADataIndices, Error> {
+        match self {
+            AppendOnlyData::PubSeq(data) => Ok(ADataIndices::new(
+                data.entry_index(),
+                data.owners_index(),
+                data.permissions_index(),
+            )),
+            AppendOnlyData::PubUnseq(data) => Ok(ADataIndices::new(
+                data.entry_index(),
+                data.owners_index(),
+                data.permissions_index(),
+            )),
+            AppendOnlyData::UnpubSeq(data) => Ok(ADataIndices::new(
+                data.entry_index(),
+                data.owners_index(),
+                data.permissions_index(),
+            )),
+            AppendOnlyData::UnpubUnseq(data) => Ok(ADataIndices::new(
+                data.entry_index(),
+                data.owners_index(),
+                data.permissions_index(),
+            )),
+        }
+    }
+
+    pub fn last_entry(&self) -> Option<(Vec<u8>, Vec<u8>)> {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.last(),
+            AppendOnlyData::PubUnseq(data) => data.last(),
+            AppendOnlyData::UnpubSeq(data) => data.last(),
+            AppendOnlyData::UnpubUnseq(data) => data.last(),
+        }
+    }
+
+    pub fn get_owners(&self, idx: u64) -> Option<&ADataOwner> {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::PubUnseq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::UnpubSeq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::UnpubUnseq(data) => data.fetch_owner_at_index(idx),
+        }
+    }
+    pub fn get_pub_user_permissions(
```

Comment:
There seem to be some formatting problems - could you please run `rustfmt` on this file?

Reply:
Hmm, I did run `rustfmt` @nbaksalyar. The CI passed too. 🤔

Reply:
Ah, could you please add a new line above this function definition to be consistent with the rest of the code style? :slightly_smiling_face: